src/cargo/ops/cargo_compile.rs

```diff
@@ -169,20 +178,44 @@ pub fn compile_ws<'a>(ws: &Workspace<'a>,
         bail!("jobs must be at least 1")
     }
 
-    let profiles = root_package.manifest().profiles();
-    if spec.len() == 0 {
-        try!(generate_targets(root_package, profiles, mode, filter, release));
-    }
+    let spec: Cow<'a, [String]> = match spec {
+        Packages::Packages(spec) => spec.into(),
+        Packages::All => ws.members()
+                           .map(|package| {
+                               let package_id = package.package_id();
+                               PackageIdSpec::from_package_id(package_id).to_string()
```

Comment:
Instead of round-tripping through `PackageIdSpec`, perhaps the resolution to a list of local packages could be done sooner?